spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/SerialShutdownHooks.scala

```diff
@@ -0,0 +1,29 @@
+package com.datastax.spark.connector.util
+
+import scala.collection.concurrent.TrieMap
+
+import org.apache.spark.Logging
+
+private[connector] object SerialShutdownHooks extends Logging {
+
+  private val hooks = new TrieMap[String, () => Unit]
+
+  def add(name: String)(body: () => Unit): Unit = {
+    hooks.put(name, body)
```

Comment:
Just add some precondition so that we can be sure we won't add anything when shutdown is in progress.